### PR TITLE
Keep query field order when using Batch Edit

### LINKED ISSUE #6597: Batch Edit columns are presented in a different order than the query
**Describe the bug**
When users build a query, there is a large degree of intentionality in how it is ordered. 
Like items are grouped together, which often communicates what data is necessary to see in what order to make reasonable decisions about batch edits that are going to be made.

If this order is not maintained, it be complex and difficult to make data moves.

Image

[**Label Query.json**](https://github.com/user-attachments/files/20657092/Label.Query.json) (in this example, on the KU Herpetology database)

When the query is run,  the results match the ordering in the query lines:

Image

Columns are in this order:
```csv
Catalog #,Family - Full Name,Genus - Full Name,Species - Full Name,Subspecies - Full Name,Determinations - Current,Type Status,Island Group,Country - Full Name,State - Full Name,County - Full Name,Locality,Locality Details - Island
```

The user is often prompted to add a bunch of tree ranks:

Image

So be it, there is no other choice at this time. The next step:

Image

Then the columns are in this order:
```csv
Catalog #,Locality,Default Geography Definition - Country - Geography name,Default Geography Definition - County - Geography name,Default Geography Definition - State - Geography name,Island Group,Locality Details - Island,Determinations - Current,Type Status,Herps taxonomy tree - Family - Taxon name,Herps taxonomy tree - Genus - Taxon name,Herps taxonomy tree - Species - Taxon name,Herps taxonomy tree - Subfamily - Taxon name,Herps taxonomy tree - Subgenus - Taxon name,Herps taxonomy tree - Subspecies - Taxon name,Herps taxonomy tree - Subtribe - Taxon name,Herps taxonomy tree - Tribe - Taxon name
```

### Side-by-side comparison

| Column name                                                                 | Pos in Query Builder | Pos in Batch Edit | Notes                                        |
|-----------------------------------------------------------------------------|:------------:|:------------:|----------------------------------------------|
| Catalog #                                                                    |      1       |      1       | identical                                    |
| Family                                                                     |      2       |     11       | renamed in Batch Edit to “Herps taxonomy tree - Family - Taxon name” |
| Genus                                                                      |      3       |     12       | renamed in Batch Edit to “Herps taxonomy tree - Genus - Taxon name” |
| Species                                                                    |      4       |     13       | renamed in Batch Edit to “Herps taxonomy tree - Species - Taxon name” |
| Subspecies                                                                 |      5       |     16       | renamed in Batch Edit to “Herps taxonomy tree - Subspecies - Taxon name” |
| Determinations - Current                                                    |      6       |      8       | identical                                    |
| Type Status                                                                 |      7       |      9       | identical                                    |
| Island Group                                                                |      8       |      6       | same name, different position                |
| Country                                                                    |      9       |      3       | renamed in Batch Edit to “Default Geography Definition - Country - Geography name” |
| State                                                                      |     10       |      5       | renamed in Batch Edit to “Default Geography Definition - State - Geography name” |
| County                                                                     |     11       |      4       | renamed in Batch Edit to “Default Geography Definition - County - Geography name” |
| Locality                                                                   |     12       |      2       | identical                                    |
| Locality Details - Island                                                  |     13       |      7       | identical                                    |

Columns only in Batch Edit (not in Query Builder):

| Column name                                                       | Pos in Batch Edit |
|-------------------------------------------------------------------|:------------:|
| Herps taxonomy tree - Subfamily - Taxon name                      |     14       |
| Herps taxonomy tree - Subgenus - Taxon name                       |     15       |
| Herps taxonomy tree - Subtribe - Taxon name                       |     17       |
| Herps taxonomy tree - Tribe - Taxon name                          |     18       |

**Summary:**
-  All original columns are present in Batch Edit, but many have been **prefixed** (e.g. “Herps taxonomy tree – …”) or **re-ordered**.
-  Batch Edit introduces four extra taxonomy-level columns under the “Herps taxonomy tree” hierarchy.
-  The “Country/State/County – Full Name” trio has been replaced with “Default Geography Definition – … – Geography name.”

**To Reproduce**

1. Build a query
2. Save it
3. Run the query
4. Click "Batch Edit"
5. See that the order is not respected

*When reporting an issue on the `production` branch or any development version of Specify, please try to recreate the issue on the latest tagged release (`v7`) and indicate if it does not occur there, indicating a regression.*

**Expected behavior**
The columns in Batch Edit should be in the **same order** as the columns in the query results.

**Crash Report**

If the bug resulted in an error message, please click on "Download Error Message" and attach it here. If there is no error, please follow [these instructions](https://discourse.specifysoftware.org/t/download-specify-7-system-information/1614) to download your System Information.

Please fill out the following information manually:
- OS: macOS
- Browser: Arc
- Specify 7 Version: `v711-prerelease` (https://github.com/specify/specify7/commit/13e71358f998e1b91dd5d3fcffa6ccb0a690e94d)
- Database Name: `KUHerps_PreservationChain`, local database
- Collection name: `Herpetology`
- User Name: `spherpsadmin`

### PR DESCRIPTION
**Pull request by alesan99**
Fixes #6597

Preserves order from query to batch edit.

Query:

image

New:
image

Old: 
image

What this PR does not do:
- Order new tree rank columns by rank id. Should this be a new issue, Or in this PR?

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [x] Add relevant issue to release milestone
- [ ] Add pr to documentation list
- [ ] Add automated tests

### Testing instructions

- Create a query with many fields, including ones from different tables.
  - Fields from different tables shouldn't get grouped together. Test this by putting fields in-between fields belonging to other tables.
  - Also test fields with the same name. Like cataloger last name and determiner last name, for example.
  - You can also test duplicate columns like having CO -> determination field and having a collection object with multiple determinations.
- Click Batch Edit
- [ ] See that the order stayed the same as in the original query.
  - If columns were duplicated, they should be grouped by record #, while still remaining in the query order
```
"Determination integer1",
"Determination remarks",
"Determination integer1 #2",
"Determination remarks #2",
"Determination integer1 #3",
"Determination remarks #3",
```
- [ ] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- Create a query with tree fields. Batch edit should prompt you to add missing ranks.
- [ ] See that the order is the same, and all new columns were added to the end.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved visual ordering in batch-edit columns: original field positions are preserved, duplicate columns keep their relative grouping, and newly visible fields are appended to the end for a more predictable layout. Tests updated to reflect the reordered headers.

[Review Change Stack](https://app.coderabbit.ai/change-stack/specify/specify7/pull/8023)

### CODE REVIEW

File: specifyweb/backend/stored_queries/batch_edit.py [hunk 1]
```diff
@@ -1095,7 +1095,24 @@ def _get_orig_column(string_id: str):
...
+    visual_order: list[int | None] = [None for _ in key_and_headers]
+    new_columns: list[int] = []
+    for index, (key, _header) in headers_enumerated:
+        # Find the column's original position if it existed in the origin query
+        original_place = key[0]
+        if original_place < len(visible_fields) and visual_order[original_place] is None:
+            visual_order[original_place] = index
+        else:
+            # New field/column. Add it to the end of the dataset.
+            new_columns.append(index)
```
[the remark this answers is not in the document]
Comment (alesan99) [reply specifyweb/backend/stored_queries/batch_edit.py]: Fields that appear multiple times still appear to be in the correct order

**Comment by alesan99** [comment]
@specify/ux-testing Should be good to test. If the PR is found to work correctly I'll update the tests.

**Review by emenslin — APPROVED** [review]
- [x] See that the order stayed the same as in the original query.
- [x] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [x] See that the order is the same, and all new columns were added to the end.

Looks good, I didn't run into any new issues

**Review by Iwantexpresso — APPROVED** [review]
- [x] See that the order stayed the same as in the original query.
- [x] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [x] See that the order is the same, and all new columns were added to the end.

Nice work!

**Comment by alesan99** [comment]
NOTE:
As an effect of preserving the original query order, new columns are grouped together by field, not by record. Should this behavior be kept @grantfitzsimmons?
The original grouping is logical but it gets tricky if the fields belonging to the table weren't grouped together in the first place (like if its `Determination integer1 | CollectionObject catalogNumber | Determination remarks`), I don't see a way cleanly  without just ignoring the query order sometimes.

Old:
```
"Determination integer1",
"Determination remarks",
"Determination integer1 #2",
"Determination remarks #2",
"Determination integer1 #3",
"Determination remarks #3",
```
New:
```
"Determination integer1",
"Determination integer1 #2",
"Determination integer1 #3",
"Determination remarks",
"Determination remarks #2",
"Determination remarks #3",
```

**Review by bhumikaguptaa — APPROVED** [review]
- [X] See that the order stayed the same as in the original query.
- [X] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [X] See that the order is the same, and all new columns were added to the end.

Works as expected!

**Comment by grantfitzsimmons** [comment]
@alesan99 My feeling is that the query field order should be preserved.

> The original grouping is logical but it gets tricky if the fields belonging to the table weren't grouped together in the first place (like if its `Determination integer1 | CollectionObject catalogNumber | Determination remarks`), I don't see a way cleanly  without just ignoring the query order sometimes.

Since the query lines show `Determination integer1 | CollectionObject catalogNumber | Determination remarks` in that order, I think it makes sense to retain the `Determination integer1` then `Determination remarks` order for each subsequent determination:

```
"Determination integer1",
"Determination remarks",
"Determination integer1 #2",
"Determination remarks #2",
"Determination integer1 #3",
"Determination remarks #3",
```

**Comment by tlammer** [comment]
@alesan99 @grantfitzsimmons The order Grant describes is definitely what the user will expect and how the user will envision their data.  It will be seen as a bug if it orders by field (integer1, integer1 #2, etc.) rather than by record.

**Comment by alesan99** [comment]
I will order the columns by record whenever there are multiple columns for a field 👍 Going to set the PR to a draft in the meanwhile

**Review by CarolineDenis — APPROVED** [review]

**Comment by alesan99** [comment]
@specify/ux-testing I have updated this PR, its ready for review again 👍

This is the new behavior for fields with duplicate columns. All other fields should still be in the query order
```
"Determination integer1",
"Determination remarks",
"Determination integer1 #2",
"Determination remarks #2",
"Determination integer1 #3",
"Determination remarks #3",
```

**Review by emenslin — CHANGES_REQUESTED** [review]
- [x] See that the order stayed the same as in the original query.
- [x] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [ ] See that the order is the same, and all new columns were added to the end.

---

When tree fields are included in the query, the batch edit results are displayed incorrectly. 

With trees: 

https://github.com/user-attachments/assets/d38e4805-ff49-4f13-bd32-302846152e7b


Query: https://ojsmnh20260326-issue-6597.test.specifysystems.org/specify/query/200/

Without trees:

https://github.com/user-attachments/assets/6ecc223f-f3ae-4846-91cc-dc6fa5670652


Query: https://ojsmnh20260326-issue-6597.test.specifysystems.org/specify/query/328/

**Review by bhumikaguptaa — CHANGES_REQUESTED** [review]
- [ ] See that the order stayed the same as in the original query.
- [ ] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [ ] See that the order is the same, and all new columns were added to the end.

While i was creating the query, if I map any field from the Project table I run into an error. And the query never loads. 

Crash report: [Specify 7 Crash Report - 2026-05-13T16_06_03.003Z.txt](https://github.com/user-attachments/files/27718784/Specify.7.Crash.Report.-.2026-05-13T16_06_03.003Z.txt)

Link to query: https://ojsmnh20260326-issue-6597-2.test.specifysystems.org/specify/query/330/

**Comment by alesan99** [comment]
@specify/ux-testing Reworked this PR, ready for testing again 👍

The following cases should work
- Columns with the same name (Cataloger "First Name", Determination Determiner "First Name")
- Columns that were duplicated in a to-many relationship ("Determiner First Name # 1", "Determiner Last Name # 1", "Determiner First Name # 2")
- New columns should still be added to the end (Like missing tree ranks)

Also updated the testing isntructions

**Review by emenslin — APPROVED** [review]
- [x] See that the order stayed the same as in the original query.
- [x] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [x] See that the order is the same, and all new columns were added to the end.

Looks good, my previous issues were fixed.

**Review by g1rly-c0d3r — COMMENTED** [review]
- [X]  See that the order stayed the same as in the original query.
       
- [X]  Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching

- [x]  See that the order is the same, and all new columns were added to the end

Works as described, columns are duplicated correctly, and tree fields are added to the end.

**Review by bhumikaguptaa — APPROVED** [review]
- [X] See that the order stayed the same as in the original query.
- [X] Make sure there are no glitches when sorting by column, looking at the mapping, editing cells, and searching
- [X] See that the order is the same, and all new columns were added to the end.

Looks good to me, there were no tables causing any errors or timeouts!

**Review by CarolineDenis — APPROVED** [review]
